src/domain_bridge/domain_bridge.cpp

```diff
@@ -0,0 +1,218 @@
+#include <string>
+#include <utility>
+
+#include "domain_bridge/exceptions.hpp"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/multi_threaded_executor.hpp"
+#include "rcutils/logging_macros.h"
+#include "rosbag2_cpp/typesupport_helpers.hpp"
+
+#include "generic_publisher.hpp"
+#include "generic_subscription.hpp"
+
+namespace domain_bridge
+{
+
+/// Implementation of \ref DomainBridge.
+class DomainBridgeImpl
+{
+public:
+  using SubscriptionMap =
+    std::map<std::pair<size_t, std::string>, std::shared_ptr<GenericSubscription>>;
+  using PublisherMap = std::map<std::pair<size_t, std::string>, std::shared_ptr<GenericPublisher>>;
+  using NodeMap = std::map<size_t, std::shared_ptr<rclcpp::Node>>;
+
+  DomainBridgeImpl()
+  {
+    this->executor_context_ = create_context_with_domain_id(1);
+    rclcpp::ExecutorOptions executor_options;
+    executor_options.context = this->executor_context_;
+    this->executor_ = std::make_unique<rclcpp::executors::MultiThreadedExecutor>(executor_options);
+  }
+
+  ~DomainBridgeImpl()
+  {
+    const std::string reason("Domain bridge shutdown");
+    this->executor_context_->shutdown(reason);
+  }
+
+  rclcpp::Context::SharedPtr create_context_with_domain_id(size_t domain_id)
+  {
+    auto context = std::make_shared<rclcpp::Context>();
+    rclcpp::InitOptions options;
+    options.auto_initialize_logging(false).set_domain_id(domain_id);
+    context->init(0, nullptr, options);
+    return context;
+  }
+
+  rclcpp::NodeOptions create_node_options(rclcpp::Context::SharedPtr context)
+  {
+    rclcpp::NodeOptions options;
+    return options.context(context)
+           .use_global_arguments(false)
+           .start_parameter_services(false)
+           .start_parameter_event_publisher(false);
+  }
+
+  rclcpp::Node::SharedPtr get_node_for_domain(size_t domain_id)
+  {
+    auto domain_id_node_pair = node_map_.find(domain_id);
+
+    // If we don't already have a node for the domain, create one
+    if (node_map_.end() == domain_id_node_pair) {
+      auto context = create_context_with_domain_id(domain_id);
+      auto node_options = create_node_options(context);
+      std::ostringstream oss;
+      oss << "domain_bridge_" << std::to_string(domain_id);
+      auto node = std::make_shared<rclcpp::Node>(oss.str(), node_options);
+      node_map_[domain_id] = node;
+      return node;
+    }
+
+    return domain_id_node_pair->second;
+  }
+
+  std::shared_ptr<GenericPublisher> create_publisher(
+    rclcpp::Node::SharedPtr node,
+    PublisherMap & map,
+    const std::string & topic_name,
+    const rosidl_message_type_support_t & typesupport_handle)
+  {
+    size_t domain_id = node->get_node_options().context()->get_domain_id();
+    auto domain_id_topic_pair = std::make_pair(domain_id, topic_name);
+    if (map.find(domain_id_topic_pair) != map.end()) {
+      throw BridgeExistsError(
+              node->get_node_options().context()->get_domain_id(),
+              topic_name
+      );
```

Comment:
I'm not sure if throwing an error here is the best thing to do

Reply:
I suppose we could print a warning instead.

Reply:
Printing a warning instead: fbabc03